sourcetracker/_cli/gibbs.py

```diff
@@ -134,96 +140,76 @@ def gibbs(table_fp, mapping_fp, output_dir, loo, jobs, alpha1, alpha2, beta,
-                    alpha1=alpha1, alpha2=alpha2, beta=beta,
-                    restarts=restarts, draws_per_restart=draws_per_restart,
-                    burnin=burnin, delay=delay,
-                    sample_metadata=sample_metadata,
-                    sources_data=sources_data, sources_envs=sources_envs,
-                    biom_table=biom_table, output_dir=output_dir)
-        sample_iter = source_samples
-    else:
-        f = partial(_cli_sink_source_prediction_runner, alpha1=alpha1,
-                    alpha2=alpha2, beta=beta, restarts=restarts,
-                    draws_per_restart=draws_per_restart, burnin=burnin,
-                    delay=delay, sources_data=sources_data,
-                    biom_table=biom_table, output_dir=output_dir)
-        sample_iter = sink_samples
+    # Prepare sink data and rarify if requested.
+    sinks = feature_table.loc[sink_samples, :]
+    if sink_rarefaction_depth > 0:
+        sinks = subsample_dataframe(sinks, sink_rarefaction_depth)
 
+    # If we've been asked to do multiple jobs, we need to spin up a cluster.
     if jobs > 1:
         # Launch the ipcluster and wait for it to come up.
         subprocess.Popen('ipcluster start -n %s --quiet' % jobs, shell=True)
         time.sleep(cluster_start_delay)
-        c = Client()
-        c[:].map(f, sample_iter, block=True)
-        # Shut the cluster down. Answer taken from SO:
-        # http://stackoverflow.com/questions/30930157/stopping-ipcluster-engines-ipython-parallel
-        c.shutdown(hub=True)
+        cluster = Client()
     else:
-        for sample in sample_iter:
-            f(sample)
-
-    # Format results for output.
-    samples = []
-    samples_data = []
-    for sample_fp in glob.glob(os.path.join(output_dir, '*')):
-        samples.append(sample_fp.strip().split('/')[-1].split('.txt')[0])
-        samples_data.append(np.loadtxt(sample_fp, delimiter='\t'))
-    mp, mps = _cli_collate_results(samples, samples_data, sources_envs)
-
-    o = open(os.path.join(output_dir, 'mixing_proportions.txt'), 'w')
-    o.writelines(mp)
-    o.close()
-    o = open(os.path.join(output_dir, 'mixing_proportions_stds.txt'), 'w')
-    o.writelines(mps)
-    o.close()
+        cluster = None
+
+    # Run the computations.
+    if not loo:
+        mpm, mps, fas = _gibbs(csources, sinks, alpha1, alpha2, beta, restarts,
+                               draws_per_restart, burnin, delay,
+                               cluster=cluster,
+                               create_feature_tables=full_output)
+    else:
+        mpm, mps, fas = _gibbs_loo(csources, alpha1, alpha2, beta, restarts,
+                                   draws_per_restart, burnin, delay,
+                                   cluster=cluster,
+                                   create_feature_tables=full_output)
+    # If we started a cluster, shut it down.
+    if jobs > 1:
+        cluster.shutdown(hub=True)
```

Comment:
Here:

https://docs.python.org/3/library/atexit.html

On Aug 22, 2016 8:46 PM, "Will Van Treuren" <notifications@github.com>
wrote:

> In sourcetracker/_cli/gibbs.py
> <https://github.com/biota/sourcetracker2/pull/51#discussion_r75798905>:
>
> > +        cluster = None
> > +
> > +    # Run the computations.
> > +    if not loo:
> > +        mpm, mps, fas = _gibbs(csources, sinks, alpha1, alpha2, beta, restarts,
> > +                               draws_per_restart, burnin, delay,
> > +                               cluster=cluster,
> > +                               create_feature_tables=full_output)
> > +    else:
> > +        mpm, mps, fas = _gibbs_loo(csources, alpha1, alpha2, beta, restarts,
> > +                                   draws_per_restart, burnin, delay,
> > +                                   cluster=cluster,
> > +                                   create_feature_tables=full_output)
> > +    # If we started a cluster, shut it down.
> > +    if jobs > 1:
> > +        cluster.shutdown(hub=True)
>
> I think its stays open - definitely should figure out how to deal with
> this. Issue #20 <https://github.com/biota/sourcetracker2/issues/20>.
> @wasade <https://github.com/wasade> - can you point me to an appropriate
> example of how to use the sys.atexit call or pattern?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/biota/sourcetracker2/pull/51/files/9649718be839d4662d8a06a592ea6a5d2d9c473f#r75798905>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAc8shrG48Q526Dk3Ra1sWPGheqbZnWiks5qim0rgaJpZM4JNBOB>
> .
>